### PR TITLE
Broadcast shouldn't use inbounds when calling f

### LINKED ISSUE #19203: Broadcasted getindex does not check bounds
```julia
julia> getindex.([[1]], 1)
1-element Array{Int64,1}:
 1

julia> getindex.([[1]], 10)
1-element Array{Int64,1}:
 0

julia> getindex.([[1]], 100000000000)

signal (11): Segmentation fault: 11
while loading no file, in expression starting on line 0
macro expansion at ./broadcast.jl:149 [inlined]
macro expansion at ./simdloop.jl:72 [inlined]
macro expansion at ./broadcast.jl:142 [inlined]
_broadcast! at ./broadcast.jl:137
unknown function (ip: 0x31810b2d7)
jl_apply_generic at /usr/local/julia-release-0.5/src/gf.c:1861
broadcast! at ./broadcast.jl:201
jl_apply_generic at /usr/local/julia-release-0.5/src/gf.c:1861
jl_apply at /usr/local/julia-release-0.5/src/./julia.h:1378 [inlined]
jl_f__apply at /usr/local/julia-release-0.5/src/builtins.c:547
broadcast_t at ./broadcast.jl:257
jl_apply_generic at /usr/local/julia-release-0.5/src/gf.c:1861
broadcast at ./broadcast.jl:18
jl_apply_generic at /usr/local/julia-release-0.5/src/gf.c:1861
do_call at /usr/local/julia-release-0.5/src/interpreter.c:71
eval at /usr/local/julia-release-0.5/src/interpreter.c:210
eval_body at /usr/local/julia-release-0.5/src/interpreter.c:480
jl_interpret_toplevel_thunk at /usr/local/julia-release-0.5/src/interpreter.c:628
jl_toplevel_eval_flex at /usr/local/julia-release-0.5/src/toplevel.c:642
jl_toplevel_eval_in at /usr/local/julia-release-0.5/src/builtins.c:577
eval at ./boot.jl:238
jlcall_eval_18352 at /usr/local/julia-release-0.5/usr/lib/julia/sys.dylib (unknown line)
jl_apply_generic at /usr/local/julia-release-0.5/src/gf.c:1861
eval_user_input at ./REPL.jl:66
unknown function (ip: 0x3180fda16)
jl_apply_generic at /usr/local/julia-release-0.5/src/gf.c:1861
macro expansion at ./REPL.jl:97 [inlined]
#3 at ./event.jl:65
unknown function (ip: 0x3180f50ef)
jl_apply_generic at /usr/local/julia-release-0.5/src/gf.c:1861
jl_apply at /usr/local/julia-release-0.5/src/./julia.h:1378 [inlined]
start_task at /usr/local/julia-release-0.5/src/task.c:259
Allocations: 3727192 (Pool: 3725866; Big: 1326); GC: 5
fish: 'julia' terminated by signal SIGSEGV (Address boundary error)
```

### PR DESCRIPTION
**Pull request by mbauman**
Previously, `broadcast(getindex, A, I...)` would call `getindex` within an
at-inbounds context. This patch simply moves the function call so that it
will throw a BoundsError appropriately.

Fixes #19203.

### CODE REVIEW

File: test/broadcast.jl [hunk 1]
```diff
@@ -337,3 +337,8 @@ end
 @test broadcast(+, 1.0, (0, -2.0)) == (1.0,-1.0)
 @test broadcast(+, 1.0, (0, -2.0), [1]) == [2.0, 0.0]
 @test broadcast(*, ["Hello"], ", ", ["World"], "!") == ["Hello, World!"]
+
+# Ensure that broadcast doesn't use @inbounds when calling the function
```
Comment (mbauman) [inline test/broadcast.jl]: The tests are run with `—check-bounds=yes`, no?  What's the best way to make sure that this catches regressions?

Comment (yuyichao) [inline test/broadcast.jl]: `test/boundscheck.jl`

**Review by simonster — APPROVED** [review]

**Comment by stevengj** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19188#issuecomment-258159314) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6ff48f5a63af0e8a19f29066fc3140be0c1001f1/7cade9f_vs_421f079/report.md). cc @jrevels

**Comment by stevengj** [comment]
There seem to be some relevant slowdowns.  It seems like we should still be able to use `@inbounds` somewhere, maybe in `_broadcast_getindex`?

**Comment by mbauman** [comment]
This is a bug-fix.  The loss in performance is entirely due to restoring bounds checks in places where they had been erroneously removed before.  In order to re-gain that performance, we'd have to add `@propagate_inbounds` annotations to the entire call chain, and then annotate the broadcast calls — if they're safe —  with `@inbounds f.(...)`.

Unfortunately, this uses `@noinline` (I assume) to ensure type stability of the inner loop, which precludes the use of `@propagate_inbounds`.

**Comment by stevengj** [comment]
But wouldn't adding
```jl
@inline _broadcast_getindex(::Any, A::AbstractArray, I) = @inbounds A[I]
```
be safe?  Here, the indices `I` are constructed by `broadcast` itself, and should always be inside the array bounds.

**Comment by mbauman** [comment]
Sure, but that's unrelated to this change.

**Comment by mbauman** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19188#issuecomment-258258062) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/aa1e61f1dc506fb67afbc07a06187b409d9e8cc3/b4bf141_vs_421f079/report.md). cc @jrevels

**Comment by tkelman** [comment]
I don't think this is relevant on release-0.5 without https://github.com/JuliaLang/julia/pull/16986

**Pull request merged**